File: leetcode/stack-queue/020_valid_parentheses.php
<?php

/*
Problem: Valid Parentheses

Link: https://leetcode.com/problems/valid-parentheses/
*/

class Solution
{
    public function isValid(string $s): bool
    {
        $stack = [];

        $mapping = [
            ')' => '(',
            ']' => '[',
            '}' => '{'
        ];

        foreach (str_split($s) as $char) {

            if (isset($mapping[$char])) {

                if (
                    empty($stack) ||
                    end($stack) !== $mapping[$char]
                ) {
                    return false;
                }

                array_pop($stack);

            } else {
                $stack[] = $char;
            }
        }

        return empty($stack);
    }
}


// Example
$s = "()[]{}";

$solution = new Solution();

var_dump($solution->isValid($s));

File: leetcode/stack-queue/020_valid_parentheses.py
"""
Problem: Valid Parentheses

Link: https://leetcode.com/problems/valid-parentheses/
"""

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        mapping = {
            ")": "(",
            "]": "[",
            "}": "{"
        }

        for char in s:

            if char in mapping:

                if not stack or stack[-1] != mapping[char]:
                    return False

                stack.pop()

            else:
                stack.append(char)

        return len(stack) == 0


# Example
s = "()[]{}"

solution = Solution()

print(solution.isValid(s))